app/models/work-item.ts

```diff
@@ -115,29 +121,49 @@ export async function getNextWorkItem(
 ): Promise<WorkItem> {
   let workItemData;
   try {
-    workItemData = await tx(WorkItem.table)
+    const subQuery =
+      tx(Job.table)
+        .select('username')
+        .join(`${WorkItem.table} as w`, `${Job.table}.jobID`, 'w.jobID')
+        .where({ 'w.status': 'ready', serviceID });
+    const userData = await tx(Job.table)
       .forUpdate()
-      .select(...tableFields, `${WorkflowStep.table}.operation`)
-      .join(WorkflowStep.table, function () {
-        this
-          .on(`${WorkflowStep.table}.stepIndex`, `${WorkItem.table}.workflowStepIndex`)
-          .on(`${WorkflowStep.table}.jobID`, `${WorkItem.table}.jobID`);
-      })
-      .join(Job.table, `${WorkItem.table}.jobID`, '=', `${Job.table}.jobID`)
-      // eslint-disable-next-line @typescript-eslint/naming-convention
-      .where({ 'work_items.serviceID': serviceID, 'work_items.status': WorkItemStatus.READY })
-      .whereIn('jobs.status', [JobStatus.RUNNING, JobStatus.ACCEPTED])
-      .orderBy([`${WorkItem.table}.id`])
+      .join(WorkItem.table, `${Job.table}.jobID`, '=', `${WorkItem.table}.jobID`)
+      .select(['username', 'serviceID', `${WorkItem.table}.serviceID`])
+      .max(`${Job.table}.updatedAt`, { as: 'm' })
+      .whereIn('username', subQuery)
+      .groupBy('username')
+      .orderBy('m', 'asc')
       .first();
+    if (userData?.username) {
+      workItemData = await tx(`${WorkItem.table} as w`)
+        .forUpdate()
+        .join(`${Job.table} as j`, 'w.jobID', 'j.jobID')
+        .join(`${WorkflowStep.table} as wf`, 'w.jobID', 'wf.jobID')
+        .select(...tableFields, 'wf.operation')
+        .whereIn('j.status', ['running', 'accepted'])
+        .where('w.status', '=', 'ready')
+        .where('w.serviceID', '=', serviceID)
+        .whereRaw('w.workflowStepIndex = wf.stepIndex')
```

Comment:
It was failing when I tried to use a normal `where`, something about trying to treat `wf.stepIndex` as a column name instead of a table alias + column.

Reply:
k, definitely add a comment in that case